fixed-hash/src/hash.rs

```diff
@@ -177,197 +211,234 @@ macro_rules! construct_hash {
+				let mut ret: $name = $name::default();
+				for i in 0..$n_bytes {
 					ret.0[i] = self.0[i] ^ rhs.0[i];
 				}
 				ret
 			}
 		}
 
 		/// Moving `BitXor`
-		impl ::core::ops::BitXor for $from {
-			type Output = $from;
+		impl $crate::core::ops::BitXor for $name {
+			type Output = $name;
 
 			fn bitxor(self, rhs: Self) -> Self::Output {
 				&self ^ &rhs
 			}
 		}
 
-		impl Default for $from {
-			fn default() -> Self { $from::new() }
+		impl Default for $name {
+			fn default() -> Self { $name::new() }
 		}
 
-		impl From<u64> for $from {
-			fn from(mut value: u64) -> $from {
-				let mut ret = $from::new();
+		impl From<u64> for $name {
+			fn from(mut value: u64) -> $name {
+				let mut ret = $name::new();
 				for i in 0..8 {
-					if i < $size {
-						ret.0[$size - i - 1] = (value & 0xff) as u8;
+					if i < $n_bytes {
+						ret.0[$n_bytes - i - 1] = (value & 0xff) as u8;
 						value >>= 8;
 					}
 				}
 				ret
 			}
 		}
 
-		impl<'a> From<&'a [u8]> for $from {
-			fn from(s: &'a [u8]) -> $from {
-				$from::from_slice(s)
+		impl<'a> From<&'a [u8]> for $name {
+			fn from(s: &'a [u8]) -> $name {
+				$name::from_slice(s)
 			}
 		}
 
-		impl_std_for_hash!($from, $size);
-		impl_heapsize_for_hash!($from);
-		impl_libc_for_hash!($from, $size);
-		impl_quickcheck_arbitrary_for_hash!($from, $size);
+		impl_std_for_hash!($name, $n_bytes);
+		impl_heapsize_for_hash!($name);
+		impl_libc_for_hash!($name, $n_bytes);
+		impl_quickcheck_arbitrary_for_hash!($name, $n_bytes);
 	}
 }
 
-/// Implements conversion to and from hash types of different sizes. Uses the
-/// last bytes, e.g. `From<H256> for H160` uses bytes 12..32
-/// CAUTION: make sure to call with correct sizes and the bigger type first or
-/// bad things will happen!
+/// Implements lossful conversions between the given types.
+///
+/// # Note
+///
+/// - Both types must be of different sizes.
+/// - Type `large_ty` must have a larger memory footprint compared to `small_ty`.
+///
+/// # Panics
+///
+/// Both `From` implementations will panic if sizes of the given types
+/// do not meet the requirements stated above.
+///
+/// # Example
+///
+/// ```
+/// #[macro_use] extern crate fixed_hash;
+/// construct_hash!{ struct H160(20); }
+/// construct_hash!{ struct H256(32); }
+/// impl_hash_conversions!(H256, H160);
+/// // now use it!
+/// # fn main() {
+/// assert_eq!(H256::from(H160::zero()), H256::zero());
+/// assert_eq!(H160::from(H256::zero()), H160::zero());
+/// # }
+/// ```
 #[macro_export]
 macro_rules! impl_hash_conversions {
-	($a: ident, $a_size: expr, $b: ident, $b_size: expr) => {
-		impl From<$b> for $a {
-			fn from(value: $b) -> $a {
-				debug_assert!($a_size > $b_size && $a_size % 2 == 0 && $b_size %2 == 0);
-				let mut ret = $a::new();
-				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(&value);
+	($large_ty:ident, $small_ty:ident) => {
+		impl From<$small_ty> for $large_ty {
+			fn from(value: $small_ty) -> $large_ty {
+				use std::mem;
```

Comment:
Something fishy is going on with this macro; on this branch I get warnings about missing std. If the conversions need to work in both `std`/`no_std` we should use `core::mem` (and sort out the implications in `lib.rs`). I also get a bunch of `type mismatch` test failures when enabling this feature.
In the PR description you mentioned "testing in isolation". Not sure what you meant by that exactly but I copied the files out of `parity-common` and fixed up `Cargo.toml` and ran the tests: same failures. Can you elaborate further on this aspect?

Reply:
Unfortunately I can't find where I have written something about "testing in isolation" and can't remember either. :(
I hope the GitHub downtime is over soon and things are syncing again correctly.

Reply:
> To verify that yourself you can try to test this crate in isolation locally.

Reply:
@pepyakin and me were trying to find a solution together.

The solution is to include the `"std"` feature in the `default` features.
This means that you have to build the crate with the cargo flag `--no-default-features` for a proper `no_std` environment. Fortunately this fixes the problems with doc tests above.